libafl_bolts/src/tuples.rs

```diff
@@ -25,21 +25,33 @@ use crate::HasLen;
 use crate::Named;
 
 /// Returns if the type `T` is equal to `U`, ignoring lifetimes.
+#[inline] // this entire call gets optimized away :)
 #[must_use]
 pub fn type_eq<T: ?Sized, U: ?Sized>() -> bool {
+    // decider struct: hold a cell (which we will update if the types are unequal) and some
+    // phantom data using a function pointer to allow for Copy to be implemented
     struct W<'a, T: ?Sized, U: ?Sized>(&'a Cell<bool>, PhantomData<fn() -> (&'a T, &'a U)>);
 
+    // default implementation: if the types are unequal, we will use the clone implementation
     impl<'a, T: ?Sized, U: ?Sized> Clone for W<'a, T, U> {
         fn clone(&self) -> Self {
+            // indicate that the types are unequal
+            // unfortunately, use of interior mutability (Cell) makes this not const-compatible
+            // not really possible to get around at this time
             self.0.set(false);
             W(self.0, self.1)
         }
     }
 
+    // specialized implementation: Copy is only implemented if the types are the same
     #[allow(clippy::mismatching_type_param_order)]
     impl<'a, T: ?Sized> Copy for W<'a, T, T> {}
 
     let detected = Cell::new(true);
+    // [].clone() is *specialized* in core.
```

Comment:
Any documentation on this?

Reply:
Like, anything official/some blog post, forum entry, docstring, ...?

Reply:
Sanity check here (implied specialization): https://doc.rust-lang.org/stable/std/clone/trait.Clone.html#how-can-i-implement-clone

You can see the specific specialization here: https://doc.rust-lang.org/stable/src/core/slice/mod.rs.html#4656-4686